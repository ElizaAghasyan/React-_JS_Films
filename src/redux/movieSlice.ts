import {createAsyncThunk, createSlice} from "@reduxjs/toolkit";
import serverApi, {movieType} from "../config/movieApi";

const initialState = {
    movies: {},
}

interface MovieType  {
    id: number;
    title: string;
    overview: string;
    poster_path: string;
    vote_average: number;
    backdrop_path: string;
    vote_count: number;
    name: string;
}


export const fetchAsyncMovies = createAsyncThunk('movies/fetchAsyncMovies', async () => {
    const params = {page: 1}
    let response = await serverApi.getMovieList(movieType.popular, {params})
        .catch(err => {
            console.log(err)
        })
    let result = await response?.data.results[Math.floor(Math.random() * response.data.results.length - 1)]
    return result;
});


export const movieSlice = createSlice({
    name: 'movies',
    initialState,
    reducers: {
        addMovies: (state, {payload}) => {
            state.movies = payload
        }
    },
    extraReducers: (builder) => {
        builder.addCase(fetchAsyncMovies.pending, (state, action) => {
            console.log("Pending...")
        })
        builder.addCase(fetchAsyncMovies.fulfilled, (state: {}, action) => {
            return {
                ...state, movies: action.payload
            }
        })
        builder.addCase(fetchAsyncMovies.rejected, (state, action) => {
            console.log("Pending...")
        })
    }
});

export const {addMovies} = movieSlice.actions;

export const getBannerMovie = (state: { movies: { movies: MovieType }}) => state.movies.movies;


export default movieSlice.reducer;
