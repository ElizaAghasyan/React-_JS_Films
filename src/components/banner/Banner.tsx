import { useState } from "react";
import { useSelector }  from "react-redux";
import TrailerModal from "../modal/TrailerModal";
import { getBannerMovie } from "../../redux/movieSlice";
import { apiConfig } from "../../config/apiConfig";
import { Rating } from "react-simple-star-rating";

import styles from './Banner.module.scss';

const Banner = () => {
    const movie = useSelector(getBannerMovie);
    const [show, setShow] = useState<boolean>(false);
    const [active, setActive] = useState<boolean>(false)
    const [rating, setRating] = useState<number | null>(movie?.vote_average);
    let background;

    const handleRating = () => setRating(rating)

    const showOverview = () => setShow(!show);

    movie?.backdrop_path ? background = apiConfig.imageURL(movie?.backdrop_path ? movie?.backdrop_path : movie?.poster_path)
        : background = null

    const setModalActive = () => setActive(!active);

    return (
        <div className={styles.banner}
            style={{
                backgroundImage: `url(${background})`,
                width: "100%"
            }}
        >
            <div className={styles.banner_content}>
                <div className={styles.banner_content_info}>
                    <h2>{movie?.title || movie?.name}</h2>
                    <ul>
                        <li>Adventure</li>
                        <li>Fantasy</li>
                        <li>Family</li>
                        <li>Comedy</li>
                    </ul>
                    <div className={styles.stars}>
                        <Rating
                            size={20}
                            fillColor='#3bb3df'
                            emptyColor='#3bb3df'
                            onClick={handleRating}
                            ratingValue={movie?.vote_count}
                        />
                        <button>{movie?.vote_average}</button>
                    </div>
                </div>
                <div className={styles.buttons}>
                    <button onClick={setModalActive}>Watch Now</button>
                    <button onClick={showOverview}>View Info</button>
                </div>
                <div className={show ? `${styles.active} ${styles.details}` : `${styles.details}`}>
                    {movie?.overview}
                </div>
                <TrailerModal active={active} id={movie?.id}/>
            </div>
        </div>
    )
}

export default Banner;
