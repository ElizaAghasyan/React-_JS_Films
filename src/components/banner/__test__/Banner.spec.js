import renderer from 'react-test-renderer';
import Banner from "../Banner";
import {render} from "../../../test-utils";
import {Provider} from "react-redux";
import {store} from "../../../redux/store";

describe('Banner Component', () => {
    test('If Banner component mounts', () => {
       let banner =  render(<Banner />)
        expect(banner).not.toBeNull()
    })

    test('Banner snapshot test', () => {
        const component = renderer.create(<Provider store={store}><Banner/></Provider>);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
