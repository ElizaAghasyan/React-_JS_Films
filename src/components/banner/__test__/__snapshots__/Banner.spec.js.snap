// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner Component Banner snapshot test 1`] = `
<div
  className="banner"
  style={
    Object {
      "backgroundImage": "url(https:://image.tmdb.org/t/p/original/undefined)",
      "width": "100%",
    }
  }
>
  <div
    className="banner_content"
  >
    <div
      className="banner_content_info"
    >
      <h2 />
      <ul>
        <li>
          Adventure
        </li>
        <li>
          Fantasy
        </li>
        <li>
          Family
        </li>
        <li>
          Comedy
        </li>
      </ul>
      <div
        className="stars"
      >
        <span
          className=""
        >
          <span
            aria-hidden="true"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#3bb3df",
                "cursor": "pointer",
                "display": "inline-flex",
                "height": 20,
                "transition": "",
                "width": 20,
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            aria-hidden="true"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#3bb3df",
                "cursor": "pointer",
                "display": "inline-flex",
                "height": 20,
                "transition": "",
                "width": 20,
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            aria-hidden="true"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#3bb3df",
                "cursor": "pointer",
                "display": "inline-flex",
                "height": 20,
                "transition": "",
                "width": 20,
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            aria-hidden="true"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#3bb3df",
                "cursor": "pointer",
                "display": "inline-flex",
                "height": 20,
                "transition": "",
                "width": 20,
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
          <span
            aria-hidden="true"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            style={
              Object {
                "color": "#3bb3df",
                "cursor": "pointer",
                "display": "inline-flex",
                "height": 20,
                "transition": "",
                "width": 20,
              }
            }
          >
            <svg
              fill="currentColor"
              height={20}
              viewBox="0 0 24 24"
              width={20}
            >
              <path
                d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
              />
            </svg>
          </span>
        </span>
        <button />
      </div>
    </div>
    <div
      className="buttons"
    >
      <button>
        Watch Now
      </button>
      <button
        onClick={[Function]}
      >
        View Info
      </button>
    </div>
    <div
      className="details"
    />
  </div>
</div>
`;
